s3.py

```diff
@@ -131,81 +128,156 @@ class S3Uploader(MirrorUploader):
-         },
-        {"key": S3_ADDRESSING_STYLE, "label": _("S3 addressing style"),
-         "type": "select",
-         "options": [
-             {
-                 "key": S3AddressingStyle.VIRTUAL.value,
-                 "label": _("Virtual")
-             },
-             {
-                 "key": S3AddressingStyle.PATH.value,
-                 "label": _("Path")
-             },
-             {
-                 "key": S3AddressingStyle.AUTO.value,
-                 "label": _("Auto")
-             }
-         ],
-         "description":
-             _("Buckets created after September 30, 2020, will support only virtual hosted-style requests. "
-               "Path-style requests will continue to be supported for buckets created on or before this date. "
-               "For more information, "
-               "see <a href='https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/'>"
-               "Amazon S3 Path Deprecation Plan - The Rest of the Story</a>."),
-         "default": S3_DEFAULT_ADDRESSING_STYLE,
-         },
-        {"key": S3_PRESIGNED_URL_EXPIRATION, "label": _("S3 presigned URL expiration"),
-         "type": "number",
-         "description": _("Time in seconds for the presigned URL to remain valid"),
-         "default": S3_DEFAULT_PRESIGNED_URL_EXPIRATION,
-         },
-        {"key": URL_TEMPLATE_KEY, "label": _("URL format"),
-         "type": "select",
-         "options": [
-             {"key": URL_TEMPLATE_DEFAULT,
-              "label": _("S3 Default: https://{bucket}.s3.{region}.amazonaws.com/{file}"),
-              },
-             {"key": URL_TEMPLATE_HTTPS,
-              "label": _("HTTPS: https://{bucket}/{file}"),
-              },
-             {"key": URL_TEMPLATE_HTTP,
-              "label": _("HTTP: http://{bucket}/{file}"),
-              },
-         ],
-         "default": URL_TEMPLATE_DEFAULT,
-         "description": _(
-             "A file mirrored to S3 is available at <code>http://{bucket}.s3.{region}.amazonaws.com/{filename}</code>. If you've set up your DNS so that http://[bucket]/ or https://[bucket]/ points to the appropriate S3 bucket, you can configure this S3 integration to shorten the URLs. <p>If you haven't set up your S3 buckets, don't change this from the default -- you'll get URLs that don't work.</p>")
-         },
+        S3UploaderConfiguration.access_key.to_settings(),
+        S3UploaderConfiguration.secret_key.to_settings(),
+        S3UploaderConfiguration.book_covers_bucket.to_settings(),
+        S3UploaderConfiguration.open_access_content_bucket.to_settings(),
+        S3UploaderConfiguration.protected_access_content_bucket.to_settings(),
+        S3UploaderConfiguration.marc_file_bucket.to_settings(),
+        S3UploaderConfiguration.s3_addressing_style.to_settings(),
+        S3UploaderConfiguration.s3_presigned_url_expiration.to_settings(),
+        S3UploaderConfiguration.url_template.to_settings()
```

Comment:
I thought the purpose of `ConfigurationMetadata.get_configuration_metadata` was to avoid needing to maintain this kind of code.

Reply:
`ConfigurationMetadata.get_configuration_metadata` returns all setting attributes and `ConfigurationBucket.to_settings` returns them in SETTINGS format. In the begging I was thinking to use just `ConfigurationBucket.to_settings` but it seems that the order in which setting attributes is not guaranteed, they somehow got mixed by `inspect.get_members`.
There are two solutions:
1. Mention all the attributes explicitly the way it's done now
2. Add `index` attribute to `ConfigurationMetadata` and sort by it

Reply:
Let's stick with #1 for now. IMO this is the biggest benefit of the whole `ConfigurationGrouping` thing so I'm a little upset that it's not working, but we can fix it later.

Reply:
I added `index` to `ConfigurationMetadata` and replaced `S3.SETTINGS` with `S3UploaderConfiguration.to_settings()`